File: main_project/tripod-new-branch/src/cpp/internalclient/ProducerManagerClient.cpp
#include <Common.h>
#include <ProducerManagerConfigClient.h>
#include <ProducerManagerClient.h>

namespace com {
namespace renren {
namespace tripod {

ProducerManagerClient::ProducerManagerClient(const std::string& zkAddress) : zkAddress_(zkAddress) {
}

DataMap ProducerManagerClient::produce(const KeySeq& keys, const std::string& businessId, bool useMaster, const long timeout) {
    std::string useWhich = "";
    std::string newUseWhich = "";
    {
        boost::shared_lock<boost::shared_mutex> lock(prxToccOrjavaMapMutex_);
        std::map<std::string, std::string>::const_iterator findIt = prxToccOrjavaMap_.find(businessId);
        if(findIt != prxToccOrjavaMap_.end()) {
            useWhich = findIt->second;
        }
    }

    DataMap res;

    if(useWhich == TRIPOD_PRODUCER_JAVA) {
        try {
            res = produceUseJava(keys, businessId, useMaster, timeout);
        }catch (TripodProducerNotSupportedException& e) {
            TRIPOD_WARN("ProducerManagerClient::produce use cc instead of java, because java not supported! businessId " + businessId);
            try {
                res = produceUseCc(keys, businessId, useMaster, timeout);
                newUseWhich = TRIPOD_PRODUCER_CC;
            }catch (Ice::Exception& e) {
                std::ostringstream os;
                os << "ProducerManagerClient::produce use cc but get Exception " << e.what() << " businessId " << businessId;
                TRIPOD_WARN(os.str());
            }
        }catch (Ice::Exception& e1) {
            std::ostringstream os;
            os << "ProducerManagerClient::produce use java but get Exception " << e1.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }

    if(useWhich == TRIPOD_PRODUCER_CC) {
        try {
            res = produceUseCc(keys, businessId, useMaster, timeout);
        }catch (TripodProducerNotSupportedException& e) {
            TRIPOD_WARN("ProducerManagerClient::produce use java instead of cc, because cc not supported! businessId " + businessId);
            try {
                res = produceUseJava(keys, businessId, useMaster, timeout);
                newUseWhich = TRIPOD_PRODUCER_JAVA;
            }catch (Ice::Exception& e) {
                std::ostringstream os;
                os << "ProducerManagerClient::produce use java but get Exception " << e.what() << " businessId " << businessId;
                TRIPOD_WARN(os.str());
            }
        }catch (Ice::Exception& e1) {
            std::ostringstream os;
            os << "ProducerManagerClient::produce use cc but get Exception " << e1.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }

    if(useWhich.empty()) {
        TRIPOD_INFO("ProducerManagerClient::produce try with cc and java businessId " + businessId);
        try {
            res = produceUseCc(keys, businessId, useMaster, timeout);
            newUseWhich = TRIPOD_PRODUCER_CC;
        }catch (TripodProducerNotSupportedException& e) {
            TRIPOD_WARN("ProducerManagerClient::produce use java instead of cc, because cc not supported! businessId " + businessId);
            try {
                res = produceUseJava(keys, businessId, useMaster, timeout);
                newUseWhich = TRIPOD_PRODUCER_JAVA;
            }catch (Ice::Exception& e) {
                std::ostringstream os;
                os << "ProducerManagerClient::produce use java but get Exception " << e.what() << " businessId " << businessId;
                TRIPOD_WARN(os.str());
            }
        }catch (Ice::Exception& e1) {
            std::ostringstream os;
            os << "ProducerManagerClient::produce use cc but get Exception " << e1.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }

    if(useWhich != newUseWhich && !newUseWhich.empty()) {
        boost::upgrade_lock<boost::shared_mutex> lock(prxToccOrjavaMapMutex_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        prxToccOrjavaMap_[businessId] = newUseWhich;
    }

    return res;
}

KeySeq ProducerManagerClient::produceKeys(const std::string& beginKey, int limit, const std::string& businessId, bool hotKey, const long timeout) {
    std::string useWhich = "";
    std::string newUseWhich = "";
    {
        boost::shared_lock<boost::shared_mutex> lock(prxToccOrjavaMapMutex_);
        std::map<std::string, std::string>::const_iterator findIt = prxToccOrjavaMap_.find(businessId);
        if(findIt != prxToccOrjavaMap_.end()) {
            useWhich = findIt->second;
        }
    }

    KeySeq res;

    if(useWhich == TRIPOD_PRODUCER_JAVA) {
        try {
            res = produceUseJava(beginKey, limit, businessId, hotKey, timeout);
        }catch (TripodProducerNotSupportedException& e) {
            TRIPOD_WARN("ProducerManagerClient::produceKeys use cc instead of java, because java not supported! businessId " + businessId);
            try {
                res = produceUseCc(beginKey, limit, businessId, hotKey, timeout);
                newUseWhich = TRIPOD_PRODUCER_CC;
            }catch (Ice::Exception& e) {
                std::ostringstream os;
                os << "ProducerManagerClient::produceKeys use cc but get Exception " << e.what() << " businessId " << businessId;
                TRIPOD_WARN(os.str());
            }
        }catch (Ice::Exception& e1) {
            std::ostringstream os;
            os << "ProducerManagerClient::produceKeys use java but get Exception " << e1.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }

    if(useWhich == TRIPOD_PRODUCER_CC) {
        try {
            res = produceUseCc(beginKey, limit, businessId, hotKey, timeout);
        }catch (TripodProducerNotSupportedException& e) {
            TRIPOD_WARN("ProducerManagerClient::produceKeys use java instead of cc, because cc not supported! businessId " + businessId);
            try {
                res = produceUseJava(beginKey, limit, businessId, hotKey, timeout);
                newUseWhich = TRIPOD_PRODUCER_JAVA;
            }catch (Ice::Exception& e) {
                std::ostringstream os;
                os << "ProducerManagerClient::produceKeys use java but get Exception " << e.what() << " businessId " << businessId;
                TRIPOD_WARN(os.str());
            }
        }catch (Ice::Exception& e1) {
            std::ostringstream os;
            os << "ProducerManagerClient::produceKeys use cc but get Exception " << e1.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }

    if(useWhich.empty()) {
        TRIPOD_INFO("ProducerManagerClient::produceKeys try with cc and java businessId " + businessId);
        try {
            res = produceUseCc(beginKey, limit, businessId, hotKey, timeout);
            newUseWhich = TRIPOD_PRODUCER_CC;
        }catch (TripodProducerNotSupportedException& e) {
            TRIPOD_WARN("ProducerManagerClient::produceKeys use java instead of cc, because cc not supported! businessId " + businessId);
            try {
                res = produceUseJava(beginKey, limit, businessId, hotKey, timeout);
                newUseWhich = TRIPOD_PRODUCER_JAVA;
            }catch (Ice::Exception& e) {
                std::ostringstream os;
                os << "ProducerManagerClient::produceKeys use java but get Exception " << e.what() << " businessId " << businessId;
                TRIPOD_WARN(os.str());
            }
        }catch (Ice::Exception& e1) {
            std::ostringstream os;
            os << "ProducerManagerClient::produceKeys use cc but get Exception " << e1.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }

    if(useWhich != newUseWhich && !newUseWhich.empty()) {
        boost::upgrade_lock<boost::shared_mutex> lock(prxToccOrjavaMapMutex_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        prxToccOrjavaMap_[businessId] = newUseWhich;
    }

    return res;
}


DataMap ProducerManagerClient::produceUseJava(const KeySeq& keys, const std::string& businessId, bool useMaster, const long timeout) {

    DataMap res;
   
    ProducerManagerConfigClientPtr config = getConfig();
    if(config == NULL) {
        return res;
    }
    
    std::map<ProducerManagerPrx, KeySeq> prxMap = config->getJavaPrxToKeysMap(keys);

    for(std::map<ProducerManagerPrx, KeySeq>::const_iterator it = prxMap.begin(); it != prxMap.end(); ++it) {
        try {
            DataMap singleRes = it->first->ice_twoway()->ice_timeout(timeout)->produce(it->second, businessId, useMaster);
            res.insert(singleRes.begin(), singleRes.end());
        }catch(Ice::Exception& e){
            std::ostringstream os;
            os << "ProducerManagerClient::produceUseJava Ice::Exception " << e.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }catch(...){
            std::ostringstream os;
            os << "ProducerManagerClient::produceUseJava UnKnownException businessId " << businessId;
            TRIPOD_WARN(os.str());
        }

    }
    return res;
}

DataMap ProducerManagerClient::produceUseCc(const KeySeq& keys, const std::string& businessId, bool useMaster, const long timeout) {
    DataMap res;
    
    ProducerManagerConfigClientPtr config = getConfig();
    if(config == NULL) {
        return res;
    }

    std::map<ProducerManagerPrx, KeySeq> prxMap = config->getCcPrxToKeysMap(keys);

    for(std::map<ProducerManagerPrx, KeySeq>::const_iterator it = prxMap.begin(); it != prxMap.end(); ++it) {
        try {
            DataMap singleRes = it->first->ice_twoway()->ice_timeout(timeout)->produce(it->second, businessId, useMaster);

            res.insert(singleRes.begin(), singleRes.end());
        }catch(Ice::Exception& e){
            std::ostringstream os;
            os << "ProducerManagerClient::produceUseCc Ice::Exception " << e.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }catch(...){
            std::ostringstream os;
            os << "ProducerManagerClient::produceUseCc UnKnownException businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }
    return res;
}

KeySeq ProducerManagerClient::produceUseJava(const std::string& beginKey, int limit, const std::string& businessId, bool hotKey, const long timeout) {

    KeySeq res;
   
    ProducerManagerConfigClientPtr config = getConfig();
    if(config == NULL) {
        return res;
    }
    
    KeySeq keys;
    keys.push_back(beginKey);
    
    std::map<ProducerManagerPrx, KeySeq> prxMap = config->getJavaPrxToKeysMap(keys);

    for(std::map<ProducerManagerPrx, KeySeq>::const_iterator it = prxMap.begin(); it != prxMap.end(); ++it) {
        try {
            KeySeq singleRes = it->first->ice_twoway()->ice_timeout(timeout)->produceKeys(beginKey, limit, businessId, hotKey);

            res.insert(res.end(), singleRes.begin(), singleRes.end());
        }catch(Ice::Exception& e){
            std::ostringstream os;
            os << "ProducerManagerClient::produceKeysUseJava Ice::Exception " << e.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }catch(...){
            std::ostringstream os;
            os << "ProducerManagerClient::produceKeysUseJava UnKnownException businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }
    return res;
}

KeySeq ProducerManagerClient::produceUseCc(const std::string& beginKey, int limit, const std::string& businessId, bool hotKey, const long timeout) {
    KeySeq res;
    
    ProducerManagerConfigClientPtr config = getConfig();
    if(config == NULL) {
        return res;
    }

    KeySeq keys;
    keys.push_back(beginKey);

    std::map<ProducerManagerPrx, KeySeq> prxMap = config->getCcPrxToKeysMap(keys);

    for(std::map<ProducerManagerPrx, KeySeq>::const_iterator it = prxMap.begin(); it != prxMap.end(); ++it) {
        try {
            KeySeq singleRes = it->first->ice_twoway()->ice_timeout(timeout)->produceKeys(beginKey, limit, businessId, hotKey);

            res.insert(res.end(), singleRes.begin(), singleRes.end());
        }catch(Ice::Exception& e){
            std::ostringstream os;
            os << "ProducerManagerClient::produceKeysUseCc Ice::Exception " << e.what() << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }catch(...){
            std::ostringstream os;
            os << "ProducerManagerClient::produceKeysUseCc UnKnownException businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }
    return res;
}

ProducerManagerConfigClientPtr ProducerManagerClient::getConfig() {
    if(config_ == NULL) {
        boost::upgrade_lock<boost::shared_mutex> lock(configMutex_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        config_ = ConfigClientFactory::getInstance()->getProducerManagerConfigClient(zkAddress_);
    }
    return config_;
}

}
}
}

File: main_project/tripod-new/src/java/TripodInternalClient/src/main/java/com/renren/tripod/internalclient/ProducerManagerClient.java
package com.renren.tripod.internalclient;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.concurrent.ConcurrentHashMap;

import com.renren.tripod.ProducerManagerPrx;
import com.renren.tripod.ProducerManagerPrxHelper;
import com.renren.tripod.TripodProducerNotSupportedException;
import com.renren.tripod.configclient.ProducerManagerConfigClient;
import com.renren.tripod.configclient.ConfigClientFactory;
import com.renren.tripod.configutil.ConfigData;
import com.xiaonei.xce.log.Tripod;

public class ProducerManagerClient {
	private ProducerManagerConfigClient _config = null;
	private String _zkAddress = null;
	private ConcurrentHashMap<String, String> _prxToccOrjavaMap = new ConcurrentHashMap<String, String>();

	public ProducerManagerClient(String zkAddress) {
		_zkAddress = zkAddress;
	}

	private ProducerManagerConfigClient getConfigClient() {
		if (_config == null) {
			synchronized (this) {
				_config = ConfigClientFactory
						.getProducerMangerConfigClient(_zkAddress);
			}
		}
		return _config;
	}

	public Map<String, byte[]> produce(List<String> keys, String businessId,
			boolean useMaster) throws TripodProducerNotSupportedException {
		return produce(keys, businessId, useMaster, 1000);
	}

	public Map<String, byte[]> produce(List<String> keys, String businessId,
			boolean useMaster, int timeout)
			throws TripodProducerNotSupportedException {
		Map<String, byte[]> res = new HashMap<String, byte[]>();
		String useWhich = _prxToccOrjavaMap.get(businessId);
		String newUseWhich = null;
		if (useWhich.isEmpty()) {
			Tripod.getLogger().info("ProducerManager.produce businessId " + businessId
					+ " useWhich is empty, useCc first");
			try {
				res = produceUseCc(keys, businessId, useMaster, timeout);
				newUseWhich = ConfigData.TRIPOD_PRODUCER_CC;
			} catch (TripodProducerNotSupportedException e) {
				Tripod.getLogger().info("ProducerManager.produce businessId " + businessId
						+ " useJava instead of useCc");
				try {
					res = produceUseJava(keys, businessId, useMaster, timeout);
					newUseWhich = ConfigData.TRIPOD_PRODUCER_JAVA;
				} catch (TripodProducerNotSupportedException e1) {
					throw e1;
				}
			}
		}

		if (useWhich.equals(ConfigData.TRIPOD_PRODUCER_CC)) {
			try {
				res = produceUseCc(keys, businessId, useMaster, timeout);
				newUseWhich = ConfigData.TRIPOD_PRODUCER_CC;
			} catch (TripodProducerNotSupportedException e) {
				Tripod.getLogger().error("ProducerManager.produce businessId "
						+ businessId + " useCc but NotSupportedException!");
				try {
					res = produceUseJava(keys, businessId, useMaster, timeout);
					newUseWhich = ConfigData.TRIPOD_PRODUCER_JAVA;
				} catch (TripodProducerNotSupportedException e1) {
					throw e1;
				}

			}
		}

		if (useWhich.equals(ConfigData.TRIPOD_PRODUCER_JAVA)) {
			try {
				res = produceUseJava(keys, businessId, useMaster, timeout);
				newUseWhich = ConfigData.TRIPOD_PRODUCER_JAVA;
			} catch (TripodProducerNotSupportedException e) {
				Tripod.getLogger().error("ProducerManager.produce businessId "
						+ businessId + " useJava but NotSupportedException!");
				try {
					res = produceUseCc(keys, businessId, useMaster, timeout);
					newUseWhich = ConfigData.TRIPOD_PRODUCER_CC;
				} catch (TripodProducerNotSupportedException e1) {
					throw e1;
				}
			}
		}
		if (!newUseWhich.isEmpty() && !useWhich.equals(newUseWhich)) {
			Tripod.getLogger().info("ProducerManager.produce businessId " + businessId
					+ " newUseWhich " + newUseWhich);
			_prxToccOrjavaMap.put(businessId, newUseWhich);
		}
		return res;
	}

	private Map<String, byte[]> produceUseCc(List<String> keys,
			String businessId, boolean useMaster, int timeout)
			throws TripodProducerNotSupportedException {
		ProducerManagerConfigClient config = getConfigClient();
		if (config == null) {
			Tripod.getLogger().error("ProducerManagerClient.produceUseCc config == null, businessId " + businessId);
			return null;
		}
		Map<ProducerManagerPrx, List<String>> prxsMap = config
				.getCCPrxToKeysMap(keys);
		Map<String, byte[]> res = new HashMap<String, byte[]>();
		for (Entry<ProducerManagerPrx, List<String>> ent : prxsMap.entrySet()) {
			ProducerManagerPrx prx = ProducerManagerPrxHelper.uncheckedCast(ent
					.getKey().ice_twoway().ice_timeout(timeout));
			Map<String, byte[]> singleRes = prx.produce(
					ent.getValue().toArray(new String[ent.getValue().size()]), businessId, useMaster);
			res.putAll(singleRes);
		}
		return res;
	}

	private Map<String, byte[]> produceUseJava(List<String> keys,
			String businessId, boolean useMaster, int timeout)
			throws TripodProducerNotSupportedException {
		ProducerManagerConfigClient config = getConfigClient();
		if (config == null) {
			Tripod.getLogger().error("ProducerManagerClient.produceUseJava config == null, businessId = " + businessId);
			return null;
		}
		Map<ProducerManagerPrx, List<String>> prxsMap = config
				.getCCPrxToKeysMap(keys);
		Map<String, byte[]> res = new HashMap<String, byte[]>();
		for (Entry<ProducerManagerPrx, List<String>> ent : prxsMap.entrySet()) {
			ProducerManagerPrx prx = ProducerManagerPrxHelper.uncheckedCast(ent
					.getKey().ice_twoway().ice_timeout(timeout));
			Map<String, byte[]> singleRes = prx.produce(
					ent.getValue().toArray(new String[ent.getValue().size()]), businessId, useMaster);
			res.putAll(singleRes);
		}
		return res;
	}

}
